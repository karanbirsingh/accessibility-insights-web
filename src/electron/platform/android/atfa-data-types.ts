// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export interface AtfaBoundingRectangle {
    'Rect.bottom': number;
    'Rect.left': number;
    'Rect.right': number;
    'Rect.top': number;
}

// Source: https://github.com/google/Accessibility-Test-Framework-for-Android/blob/master/src/main/java/com/google/android/apps/common/testing/accessibility/framework/replacements/SpannableString.java
export interface SpannableString {
    'SpannableString.rawString': string;
}

// Source: https://github.com/google/Accessibility-Test-Framework-for-Android/blob/master/src/main/java/com/google/android/apps/common/testing/accessibility/framework/uielement/ViewHierarchyElement.java
export interface ViewHierarchyElement {
    'ViewHierarchyElement.accessibilityClassName': string;
    'ViewHierarchyElement.id': number;
    'ViewHierarchyElement.className': string;
    'ViewHierarchyElement.boundsInScreen'?: AtfaBoundingRectangle;
    'ViewHierarchyElement.contentDescription'?: SpannableString;
    'ViewHierarchyElement.text'?: SpannableString;
}

// Source: https://github.com/google/Accessibility-Test-Framework-for-Android/blob/master/src/main/java/com/google/android/apps/common/testing/accessibility/framework/AccessibilityHierarchyCheckResult.java
export interface AccessibilityHierarchyCheckResult {
    'AccessibilityHierarchyCheckResult.element': ViewHierarchyElement;
    'AccessibilityHierarchyCheckResult.resultId': number;
    'AccessibilityCheckResult.checkClass': string;
    'AccessibilityCheckResult.type': string; // TODO: make an enum?
    'AccessibilityHierarchyCheckResult.metadata'?: any;
}
