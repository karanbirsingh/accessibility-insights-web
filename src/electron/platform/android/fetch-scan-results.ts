// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import axios from 'axios';
import { FeatureFlagStore } from 'background/stores/global/feature-flag-store';
import { UnifiedFeatureFlags } from 'electron/common/unified-feature-flags';
import { AndroidScanResults } from './android-scan-results';

export type ScanResultsFetcher = (port: number) => Promise<AndroidScanResults>;

export type HttpGet = typeof axios.get;

export const createScanResultsFetcher = (
    httpGet: HttpGet,
    featureFlagStore: FeatureFlagStore,
): ScanResultsFetcher => {
    return async (port: number) => {
        const featureFlagStoreData = featureFlagStore.getState();
        let resultsVersion = ''; // API for v1 results doesn't have an explicit version number
        if (featureFlagStoreData[UnifiedFeatureFlags.atfaResults]) {
            resultsVersion = '_v2';
        }

        const response = await httpGet(
            `http://localhost:${port}/AccessibilityInsights/result${resultsVersion}`,
        );
        return new AndroidScanResults(response.data);
    };
};
