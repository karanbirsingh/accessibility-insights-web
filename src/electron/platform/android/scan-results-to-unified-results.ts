// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import { AndroidScanResults } from './android-scan-results';
import { RuleInformationProviderType } from './rule-information-provider-type';

export type ConvertScanResultsToUnifiedResultsDelegate = (
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
) => UnifiedResult[];

export function convertScanResultsToUnifiedResults(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
    converters: ConvertScanResultsToUnifiedResultsDelegate[],
): UnifiedResult[] {
    const unifiedResults: UnifiedResult[] = [];

    converters?.forEach(converter => {
        unifiedResults.push(...converter(scanResults, ruleInformationProvider, uuidGenerator));
    });

    return unifiedResults;
}
