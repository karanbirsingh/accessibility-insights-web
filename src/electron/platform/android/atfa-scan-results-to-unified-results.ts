// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import {
    AndroidScanResults,
    BoundingRectangle,
    RuleResultsData,
} from 'electron/platform/android/android-scan-results';
import {
    AccessibilityHierarchyCheckResult,
    AtfaBoundingRectangle,
    SpannableString,
    ViewHierarchyElement,
} from 'electron/platform/android/atfa-data-types';
import { RuleInformation } from 'electron/platform/android/rule-information';
import { RuleInformationProviderType } from 'electron/platform/android/rule-information-provider-type';

const includedResults = ['ERROR', 'WARNING'];

export function convertAtfaScanResultsToUnifiedResults(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
): UnifiedResult[] {
    if (!scanResults || !scanResults.ruleResults) {
        return [];
    }

    const unifiedResults: UnifiedResult[] = [];

    for (const atfaResult of scanResults.atfaResults) {
        if (includeBasedOnResultType(atfaResult)) {
            const viewElement: ViewHierarchyElement =
                atfaResult['AccessibilityHierarchyCheckResult.element'];
            if (viewElement) {
                const ruleInformation: RuleInformation = ruleInformationProvider.getRuleInformation(
                    atfaResult['AccessibilityCheckResult.checkClass'],
                );

                if (ruleInformation) {
                    unifiedResults.push(
                        createUnifiedResult(
                            ruleInformation,
                            atfaResult,
                            viewElement,
                            uuidGenerator,
                        ),
                    );
                }
            }
        }
    }

    return unifiedResults;
}

function createUnifiedResult(
    ruleInformation: RuleInformation,
    atfaResult: AccessibilityHierarchyCheckResult,
    viewElement: ViewHierarchyElement,
    uuidGenerator: UUIDGenerator,
): UnifiedResult {
    const ruleResult: RuleResultsData = {
        ruleId: ruleInformation.ruleId,
        status: atfaResult['AccessibilityCheckResult.type'],
        props: atfaResult['AccessibilityHierarchyCheckResult.metadata'],
    };
    return {
        uid: uuidGenerator(),
        ruleId: ruleInformation.ruleId,
        status: ruleInformation.getResultStatus(ruleResult),
        descriptors: {
            className: viewElement?.['ViewHierarchyElement.className'],
            boundingRectangle: convertBoundingRectangle(
                viewElement?.['ViewHierarchyElement.boundsInScreen'],
            ),
            contentDescription: getRawString(
                viewElement?.['ViewHierarchyElement.contentDescription'],
            ),
            text: getRawString(viewElement?.['ViewHierarchyElement.text']),
        },
        identifiers: {
            identifier: viewElement?.['ViewHierarchyElement.accessibilityClassName'],
            conciseName: viewElement?.['ViewHierarchyElement.className'],
        },
        resolution: ruleInformation.getUnifiedResolution(ruleResult),
    };
}

function getRawString(spannableString?: SpannableString): string | null {
    if (spannableString) {
        return spannableString['SpannableString.rawString'] ?? null;
    }
    return null;
}

function includeBasedOnResultType(atfaResult: AccessibilityHierarchyCheckResult): boolean {
    const resultType: string | null = atfaResult['AccessibilityCheckResult.type'] ?? null;

    return includedResults.includes(resultType);
}

function convertBoundingRectangle(
    boundingRectangle?: AtfaBoundingRectangle,
): BoundingRectangle | undefined {
    if (!boundingRectangle) {
        return undefined;
    }

    return {
        bottom: boundingRectangle['Rect.bottom'],
        left: boundingRectangle['Rect.left'],
        right: boundingRectangle['Rect.right'],
        top: boundingRectangle['Rect.top'],
    };
}
