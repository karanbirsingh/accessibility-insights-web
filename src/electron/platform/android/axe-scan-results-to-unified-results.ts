// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import {
    AndroidScanResults,
    AxeRuleResultsData,
    ViewElementData,
} from 'electron/platform/android/android-scan-results';
import { RuleInformation } from 'electron/platform/android/rule-information';
import { RuleInformationProviderType } from 'electron/platform/android/rule-information-provider-type';
import { DictionaryStringTo } from 'types/common-types';

export function convertAxeScanResultsToUnifiedResults(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
): UnifiedResult[] {
    if (!scanResults || !scanResults.ruleResults) {
        return [];
    }

    const viewElementLookup: DictionaryStringTo<ViewElementData> =
        createViewElementLookup(scanResults);
    const unifiedResults: UnifiedResult[] = [];

    for (const ruleResult of scanResults.ruleResults) {
        const ruleInformation: RuleInformation = ruleInformationProvider.getRuleInformation(
            ruleResult.ruleId,
        );

        if (ruleInformation) {
            unifiedResults.push(
                createUnifiedResult(ruleInformation, ruleResult, viewElementLookup, uuidGenerator),
            );
        }
    }

    return unifiedResults;
}

function createViewElementLookup(
    scanResults: AndroidScanResults,
): DictionaryStringTo<ViewElementData> {
    const viewElementLookup = {};

    addViewElementAndChildren(viewElementLookup, scanResults.viewElementTree);

    return viewElementLookup;
}

function addViewElementAndChildren(
    viewElementLookup: DictionaryStringTo<ViewElementData>,
    element: ViewElementData | null,
): void {
    if (element) {
        viewElementLookup[element.axeViewId] = element;
        if (element.children) {
            for (const child of element.children) {
                addViewElementAndChildren(viewElementLookup, child);
            }
        }
    }
}

function createUnifiedResult(
    ruleInformation: RuleInformation,
    ruleResult: AxeRuleResultsData,
    viewElementLookup: DictionaryStringTo<ViewElementData>,
    uuidGenerator: UUIDGenerator,
): UnifiedResult {
    const viewElement = viewElementLookup[ruleResult.axeViewId];
    return {
        uid: uuidGenerator(),
        ruleId: ruleInformation.ruleId,
        status: ruleInformation.getResultStatus(ruleResult),
        descriptors: {
            className: viewElement?.className,
            boundingRectangle: viewElement?.boundsInScreen,
            contentDescription: viewElement?.contentDescription,
            text: viewElement?.text,
        },
        identifiers: {
            identifier: viewElement?.className,
            conciseName: viewElement?.className,
        },
        resolution: ruleInformation.getUnifiedResolution(ruleResult),
    };
}
