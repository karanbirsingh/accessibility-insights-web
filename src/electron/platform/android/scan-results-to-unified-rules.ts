// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UnifiedRule } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import { RuleInformation } from 'electron/platform/android/rule-information';
import { AndroidScanResults } from './android-scan-results';
import { RuleInformationProviderType } from './rule-information-provider-type';

export type ConvertScanResultsToUnifiedRulesDelegate = (
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
) => UnifiedRule[];

export function convertScanResultsToUnifiedRules(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
    converters: ConvertScanResultsToUnifiedRulesDelegate[],
): UnifiedRule[] {
    const unifiedRules: UnifiedRule[] = [];

    converters?.forEach(converter => {
        unifiedRules.push(...converter(scanResults, ruleInformationProvider, uuidGenerator));
    });

    return unifiedRules;
}

export function createUnifiedRuleFromRuleResult(
    ruleInformation: RuleInformation,
    uuidGenerator: UUIDGenerator,
): UnifiedRule {
    return {
        uid: uuidGenerator(),
        id: ruleInformation.ruleId,
        description: ruleInformation.ruleDescription,
        url: ruleInformation.ruleLink,
        guidance: ruleInformation.guidance,
    };
}
