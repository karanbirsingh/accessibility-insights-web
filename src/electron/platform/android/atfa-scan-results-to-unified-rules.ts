// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UnifiedRule } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import { createUnifiedRuleFromRuleResult } from 'electron/platform/android/scan-results-to-unified-rules';
import { AndroidScanResults } from './android-scan-results';
import { RuleInformationProviderType } from './rule-information-provider-type';

export type ConvertScanResultsToUnifiedRulesDelegate = (
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
) => UnifiedRule[];

export function convertAtfaScanResultsToUnifiedRules(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
): UnifiedRule[] {
    if (!scanResults) {
        return [];
    }

    const unifiedRules: UnifiedRule[] = [];
    const ruleIds: Set<string> = new Set();

    for (const atfaResult of scanResults.atfaResults) {
        const ruleId: string = atfaResult['AccessibilityCheckResult.checkClass'];
        if (!ruleIds.has(ruleId)) {
            const ruleInformation = ruleInformationProvider.getRuleInformation(ruleId);

            if (ruleInformation) {
                unifiedRules.push(createUnifiedRuleFromRuleResult(ruleInformation, uuidGenerator));
                ruleIds.add(ruleId);
            }
        }
    }

    return unifiedRules;
}
