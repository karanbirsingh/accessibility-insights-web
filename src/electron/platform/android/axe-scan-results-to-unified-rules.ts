// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { UnifiedRule } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import { AndroidScanResults } from 'electron/platform/android/android-scan-results';
import { RuleInformationProviderType } from 'electron/platform/android/rule-information-provider-type';
import { createUnifiedRuleFromRuleResult } from 'electron/platform/android/scan-results-to-unified-rules';

export function convertAxeScanResultsToUnifiedRules(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
): UnifiedRule[] {
    if (!scanResults) {
        return [];
    }

    const unifiedRules: UnifiedRule[] = [];
    const ruleIds: Set<string> = new Set();

    for (const result of scanResults.ruleResults) {
        const ruleId = result.ruleId;
        if (!ruleIds.has(ruleId)) {
            const ruleInformation = ruleInformationProvider.getRuleInformation(ruleId);

            if (ruleInformation) {
                unifiedRules.push(createUnifiedRuleFromRuleResult(ruleInformation, uuidGenerator));
                ruleIds.add(ruleId);
            }
        }
    }

    return unifiedRules;
}
