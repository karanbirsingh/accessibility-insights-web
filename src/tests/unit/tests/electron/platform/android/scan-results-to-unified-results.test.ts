// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { generateUID, UUIDGenerator } from 'common/uid-generator';
import { AndroidScanResults } from 'electron/platform/android/android-scan-results';
import { RuleInformationProvider } from 'electron/platform/android/rule-information-provider';
import { RuleInformationProviderType } from 'electron/platform/android/rule-information-provider-type';
import { convertScanResultsToUnifiedResults } from 'electron/platform/android/scan-results-to-unified-results';
import { IMock, Mock, MockBehavior } from 'typemoq';

describe('ScanResultsToUnifiesRules', () => {
    const ruleInformationProviderMock: IMock<RuleInformationProviderType> =
        Mock.ofType<RuleInformationProviderType>(undefined, MockBehavior.Strict);
    const uuidGeneratorMock: IMock<UUIDGenerator> = Mock.ofInstance(
        generateUID,
        MockBehavior.Strict,
    );
    const scanResults = {} as AndroidScanResults;

    it('returns an empty set if null converter list is specified', () => {
        const convertedRules: UnifiedResult[] = convertScanResultsToUnifiedResults(
            scanResults,
            ruleInformationProviderMock.object,
            uuidGeneratorMock.object,
            null,
        );
        expect(convertedRules.length).toBe(0);
    });

    it('returns an empty set if an empty converter list is specified', () => {
        const convertedRules: UnifiedResult[] = convertScanResultsToUnifiedResults(
            scanResults,
            ruleInformationProviderMock.object,
            uuidGeneratorMock.object,
            [],
        );
        expect(convertedRules.length).toBe(0);
    });

    it('calls each specified converter in turn and concatenates results', () => {
        const converter1 = (
            results: AndroidScanResults,
            provider: RuleInformationProvider,
            uuidGenerator: UUIDGenerator,
        ): UnifiedResult[] => {
            expect(results).toBe(scanResults);
            expect(provider).toBe(ruleInformationProviderMock.object);
            expect(uuidGenerator).toBe(uuidGeneratorMock.object);
            return [{ placeholder: 0 } as unknown as UnifiedResult];
        };
        const converter2 = (
            results: AndroidScanResults,
            provider: RuleInformationProvider,
            uuidGenerator: UUIDGenerator,
        ): UnifiedResult[] => {
            expect(results).toBe(scanResults);
            expect(provider).toBe(ruleInformationProviderMock.object);
            expect(uuidGenerator).toBe(uuidGeneratorMock.object);
            return [
                { placeholder: 1 } as unknown as UnifiedResult,
                { placeholder: 2 } as unknown as UnifiedResult,
            ];
        };

        const unifiedResults: UnifiedResult[] = convertScanResultsToUnifiedResults(
            scanResults,
            ruleInformationProviderMock.object,
            uuidGeneratorMock.object,
            [converter1, converter2],
        );

        expect(unifiedResults).toMatchSnapshot();
    });
});
