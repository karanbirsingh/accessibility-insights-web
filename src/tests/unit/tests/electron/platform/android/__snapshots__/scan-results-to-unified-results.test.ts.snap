// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanResultsToUnifiesRules calls each specified converter in turn and concatenates results 1`] = `
Array [
  Object {
    "placeholder": 0,
  },
  Object {
    "placeholder": 1,
  },
  Object {
    "placeholder": 2,
  },
]
`;
